src/flint/fmpz_factor.jl

```diff
@@ -18,30 +18,45 @@ function _factor(a::ZZRingElem)
 end
 
 # Just to give a helpful error message if someone tries to factor a boolean
-function factor(b::Bool)
-  throw(DomainError("Cannot factorize a boolean"));
+function factor(::Bool)
+  throw(DomainError("Cannot factorize a boolean"))
 end
 
-# This function handles machine integer types (up to 64 bits)
-function factor(a::T) where T <: Union{Int, UInt}
+function factor(::Type{Bool}, ::Int)
+  throw(DomainError("Cannot have a factorization into booleans"))
+end
+
+function factor(::Type{Bool}, ::ZZRingElem)
+  throw(DomainError("Cannot have a factorization into booleans"))
+end
+
+
+# The main dispatch function: flintify gives either Int or ZZRingElem,
+# then use Julia's dispatcher to select the worker function.
+function factor(a::T)  where T <: Integer
+  iszero(a) && throw(ArgumentError("Argument must be non-zero"))
+  return factor(typeof(a), Nemo.flintify(a))
+end
+
+# Two worker functions: one for Int, one for ZZRingElem
+function factor(::Type{T}, a::Int)  where T <: Int
   iszero(a) && throw(ArgumentError("Argument must be non-zero"))
-  u = sign(a)
-  a = u < 0 ? -a : a
+  u = T(sign(a))
+  abs_a = reinterpret(UInt, a < 0 ? -a : a) # supposedly correct when a is most negative Int value
```

Comment:
"supposedly" ? Let's walk through it: if `a` is positive then `a < 0 ? -a : a` is positive and reinterpreting a positive `Int` to `UInt` works as intended.

If `a` is negative and *not* equal to `typemin(Int)` then `a < 0 ? -a : a` evaluates to `-a` which then is positive, and so once again the `reinterpret` is correct.

This leaves the case were `a == typemin(Int)`. In that case `-a == a` and this is negative. But this is a single value and we can e.g. verify manually that we get the correct result:
```
julia> a = typemin(Int)
-9223372036854775808

julia> -a
-9223372036854775808

julia> reinterpret(UInt, -a)
0x8000000000000000

julia> big(reinterpret(UInt, -a))
9223372036854775808

julia> big(a)
-9223372036854775808

julia> -big(a)
9223372036854775808

julia> -big(a) == big(reinterpret(UInt, -a))
true
```

So I suggest this:
```suggestion
  abs_a = reinterpret(UInt, a < 0 ? -a : a) # abs(a) fails if a is typemin(Int)
```